lib/src/kvs/ds.rs

```diff
@@ -821,25 +861,126 @@ impl Datastore {
 		// TODO Add Node GC?
 		Ok(())
 	}
 
 	// save_timestamp_for_versionstamp saves the current timestamp for the each database's current versionstamp.
-	pub(crate) async fn save_timestamp_for_versionstamp(&self, ts: u64) -> Result<(), Error> {
+	pub(crate) async fn save_timestamp_for_versionstamp(
+		&self,
+		ts: u64,
+	) -> Result<Option<Versionstamp>, Error> {
 		let mut tx = self.transaction(Write, Optimistic).await?;
-		if let Err(e) = self.save_timestamp_for_versionstamp_impl(ts, &mut tx).await {
-			return match tx.cancel().await {
-				Ok(_) => {
-					Err(e)
+		match self.save_timestamp_for_versionstamp_impl(ts, &mut tx).await {
+			Ok(vs) => Ok(vs),
+			Err(e) => {
+				match tx.cancel().await {
+					Ok(_) => {
+						Err(e)
+					}
+					Err(txe) => {
+						Err(Error::Tx(format!("Error saving timestamp for versionstamp: {:?} and error cancelling transaction: {:?}", e, txe)))
+					}
 				}
-				Err(txe) => {
-					Err(Error::Tx(format!("Error saving timestamp for versionstamp: {:?} and error cancelling transaction: {:?}", e, txe)))
+			}
+		}
+	}
+
+	/// This is a future that is from whatever is running the datastore as a SurrealDB instance (api WASM and native)
+	/// It's responsibility is to catch up all live queries based on changes to the relevant change feeds,
+	/// and send notifications after assessing authorisation. Live queries then have their watermarks updated.
+	pub async fn process_lq_notifications(&self) -> Result<(), Error> {
+		// Runtime feature gate, as it is not production-ready
+		if !FFLAGS.change_feed_live_queries.enabled() {
+			return Ok(());
+		}
+		// Return if there are no live queries
+		if self.notification_channel.is_none() {
+			return Ok(());
+		}
+		if self.local_live_queries.read().await.is_empty() {
+			return Ok(());
+		}
+
+		// Find live queries that need to catch up
+		let mut change_map: BTreeMap<LqSelector, Vec<ChangeSet>> = BTreeMap::new();
+		let mut tx = self.transaction(Read, Optimistic).await?;
+		for (selector, vs) in self.local_live_query_cfs.read().await.iter() {
+			// Read the change feed for the selector
+			let res = cf::read(
+				&mut tx,
+				&selector.ns,
+				&selector.db,
+				// Technically, we can not fetch by table and do the per-table filtering this side.
+				// That is an improvement though
+				Some(&selector.tb),
+				ShowSince::versionstamp(vs),
+				Some(TEMPORARY_LQ_CF_BATCH_SIZE_TILL_WE_HAVE_PAGINATION),
+			)
+			.await;
+			let changes = res.map(|change_sets| {
+				let new_watermark = change_sets.last().map(|change_set| {
+					conv::versionstamp_to_u64({
+						let v: &Versionstamp = &change_set.0;
+						v
+					})
+				});
+
+				// Confirm we do need to change watermark - this is technically already handled by the cf range scan
+				match new_watermark {
+					Some(latest_vs) => {
+						if latest_vs > conv::versionstamp_to_u64(vs) {
+							Some(change_sets)
+						} else {
+							None
+						}
+					}
+					None => None,
 				}
+			})?;
```

Comment:
I think this here is a bit complicated with multiple options returned, I think it could be simplified to the following:

```suggestion
			let change_sets = cf::read(
				&mut tx,
				&selector.ns,
				&selector.db,
				// Technically, we can not fetch by table and do the per-table filtering this side.
				// That is an improvement though
				Some(&selector.tb),
				ShowSince::versionstamp(vs),
				Some(TEMPORARY_LQ_CF_BATCH_SIZE_TILL_WE_HAVE_PAGINATION),
			)
			.await?;
			// Confirm we do need to change watermark - this is technically already handled by the cf range scan
			if let Some(change_set) = change_sets.last() {
				if conv::versionstamp_to_u64(&change_set.0) > conv::versionstamp_to_u64(vs) {
					change_map.insert(selector.clone(), change_sets);
				}
			}
```